proposals/collection-expressions-inline.md

```diff
@@ -0,0 +1,124 @@
+# Collection expressions: inline collections
+
+## Summary
+
+Support collection expressions *inline* in expression contexts where the *collection type* is not observable.
+
+## Motivation
+
+Inline collection expressions could be used with spreads to allow adding elements *conditionally* to the containing collection:
+```csharp
+int[] items = [x, y, .. b ? [z] : []];
+```
+
+Inline collection expressions could be used directly in `foreach`:
+```csharp
+foreach (var b in [false, true]) { }
+```
+In these cases, the *collection type* of the inline collection expression is unspecified. The choice of how or whether to instantiate the collection is left to the compiler.
+
+## Detailed design
+
+### Conversions
+
+To support conversions for inline collection expressions, an abstract *collection_type* is introduced.
+
+A *collection_type* represents an enumerable type with a specific element type and an *unspecified* collection type.
+A *collection_type* with element type `E` is referred to here as *`col<E>`*.
+
+A *collection_type* exists at compile time only; a *collection_type* cannot be referenced in source or metadata.
+
+The [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) of a *collection_type* `col<E>` is `E`.
+
+An implicit *collection_type conversion* exists from a type with an [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) `Tₑ` to the *collection_type* `col<Tₑ>`.
+
+The collection expression [*conversions*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#conversions) section is updated as follows:
+
+> An implicit *collection expression conversion* exists from a collection expression to the following types:
+> * **A *collection_type* `col<T>`**
+> * A single dimensional *array type* `T[]`
+> * ...
+> 
+> The implicit *collection expression conversion* exists if the type has an [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) `Tₑ` where for each *element* `Eᵢ` in the collection expression:
+> * If `Eᵢ` is an *expression element*, there is an implicit conversion from `Eᵢ` to `Tₑ`.
+> * If `Eᵢ` is a *spread element* `..Sᵢ`, there is an implicit conversion **from `Sᵢ` to the *collection_type* `col<Tₑ>`**.
+
+### Construction
+
+The elements of a nested collection expression within a spread can be added to the containing collection instance directly, without instantiating an intermediate collection.
+
+The elements of the nested collection expression are evaluated in order within the nested collection. There is no implied evaluation order between the elements of the nested collection and other elements within the containing collection expression.
```

Comment:
this seems wrong to me.  there should def be an implied evaluation order.  if i have 

```[a, b, .. c ? [d] : [e], f, g]```

i would expect that `d` and `e` would be evaluated before `f` and `g`.

Reply:
Changed to:
> The elements of a collection expression within a spread are evaluated in order as if the spread elements were declared in the containing collection expression directly.